### ISSUE TITLE
Sanic no longer supports awaitables for error handling

### ISSUE THREAD

**Issue by autumnjolitz** [ashleysommer/sanic-cors#38]
This particular line really sent my day for some delving in to Sanic:

https://github.com/ashleysommer/sanic-cors/blob/b4fc0b2d83d3c4f596ff85ff07a84abb0003c516/sanic_cors/extension.py#L326-L355

Basically, this line (https://github.com/huge-success/sanic/blob/927c0e082e0fe22e33c00395dfcfc7b6fdf0b7dd/sanic/server.py#L541) *never* supported an `awaitable` to be returned from `Response.output(...)` (synchronous function).

**Comment by benjolitz** [comment]
Had to detect it by overriding `sanic.server.HttpProtocol.write_error` with:
```
from sanic import server
def write_error(self, exception):
    response = self.error_handler.response(self.request, exception)
    if inspect.isawaitable(response):
        logger.exception(f'wtf! - {self.error_handler}', exc_info=exception)
    return _write_error(self, exception)

server.HttpProtocol.write_error = write_error```

Symptoms:
```
api_1       | Traceback (most recent call last):
api_1       |   File "/usr/local/lib/python3.7/site-packages/sanic/server.py", line 540, in write_error
api_1       |     self.transport.write(response.output(version))
api_1       | AttributeError: 'coroutine' object has no attribute 'output'```

**Comment by ashleysommer** [comment]
Hi @benjolitz 
Is this a new issue that appeared 20 hours ago after merging https://github.com/ashleysommer/sanic-cors/commit/473f2cba75cd9a5f2afb380c226a2792da4662de?

**Comment by benjolitz** [comment]
That looks like it. The monkeypatch I'm using pretty much is the original code before that commit.

**Comment by ashleysommer** [comment]
Ok, I'll revert that commit. 
I will have to dive deeper into determining the actual solution here.
Some users say the exception wrapper _must_ be async, so they can use their async error handlers.
But you've just pointed out that the exception wrapper _cannot_ be async, because Sanic writes it directly to the transport output.

**Comment by benjolitz** [comment]
Weirdly enough, I had the exact same mistaken impression about async/non-async error handlers several months ago and learned the hard way that it's only synchronous. 

Perhaps the sanic docs are async-everywhere except for that one edge case of ErrorHandler.  Thanks for reverting the commit.

**Comment by ashleysommer** [comment]
@autumnjolitz 
I've finally had time to look into this deeper.
So Sanic ErrorHandlers _can_ be async. The problem was the patch I merged to add that support back in September was incomplete.
Both the exception wrapper, _and_ the Exception handler `.response()` function must be async in order for the sanic response handler to await the response.
This fix to support async error handlers will be in the next released version.

**Comment by ashleysommer** [comment]
Fixed by https://github.com/ashleysommer/sanic-cors/commit/2ad2bb0b699827ac6a63d985b8db8b1362977604

**Comment by autumnjolitz** [comment]
So this issue reared it's head again, despite using either the latest sanic or version 19.9.0.

I don't have any custom classes for `app.error_handler` - it's the default ErrorHandler.

Did you test the case of `raise CustomException` and handle exception in ```@app.exception(CustomException)
def handle_exc(request, exception):
    return text('error')
``` ?


I don't see how `write_error` can tolerate an async return, even with https://github.com/ashleysommer/sanic-cors/commit/2ad2bb0b699827ac6a63d985b8db8b1362977604 factored in:

Clearly sync here: https://github.com/huge-success/sanic/blob/91f6abaa81248189fbcbdf685e8bdcbb7846609f/sanic/server.py#L577

Base ErrorHandler class also sync: https://github.com/huge-success/sanic/blob/91f6abaa81248189fbcbdf685e8bdcbb7846609f/sanic/handlers.py#L74

Reverting to sanic-cors to 0.9.9 handled the issue for me.

**Comment by ashleysommer** [comment]
Hi @autumnjolitz 
I was following the logic seen here: https://github.com/huge-success/sanic/blob/91f6abaa81248189fbcbdf685e8bdcbb7846609f/sanic/app.py#L1007
In that section of Sanic, it appears the error handler is _can_ be async.

And the testing I performed was around catching exceptions in the app handle_request() method.

But you're right, when Sanic is gets an error in the lower-level asyncio server code, either before it is handling the request, or after it is handling the request, then in that case the error handler cannot be async.

I'll revert the sanic-cors async handler change for now, and push a new postfix release, and I'll discuss this inconsistency with the Sanic Core dev team.

**Comment by ashleysommer** [comment]
Fixed again in 0.10.0.post2 release
